packages/fxa-shared/metrics/validate.ts

```diff
@@ -0,0 +1,379 @@
+        0
+      );
+    };
+
+    clamp('connectStart');
+    clamp('connectEnd');
+    clamp('domainLookupEnd');
+    clamp('redirectStart');
+
+    return data;
+  }
+
+  /**
+   * There are appear to be some issues with event data. This routine ensures
+   * the event offset is a sane number [0, maxEventOffset], and that the event
+   * types are valid string values.
+   *
+   * Coercion of data will be reported and counted as a critical error.
+   */
+  public sanitizeEvents(data: Pick<CheckedMetrics, 'events'>) {
+    if (!Array.isArray(data.events)) {
+      data.events = [];
+    }
+
+    for (const event of data.events) {
+      // Make sure offset is a valid value
+      event.offset = this.clamp(
+        'event.offset',
+        event.offset,
+        0,
+        this.maxEventOffset,
+        0
+      );
+
+      // Truncate as soon as an invalid character is hit. There is a limited set of valid characters for event type;
+      // however, the invoking code is sometimes a bit fast and loose with what gets provided here. e.g. Sometimes a
+      // noisy error messages is provided. Looking at existing data, this approach should allow most scenarios to
+      // result in a distinct event.type value, which is the desired end result.
+      if (typeof event.type === 'string') {
+        event.type = event.type.replace(/([\w\s.:-]*)(.*)/, '$1');
+      }
+    }
+
+    return data;
+  }
+
+  /**
+   * Since RPs and client do not have full control over the UTM
+   * params that could be passed to them, we need to sanitize those values
+   * the best we can. This replaces and reports any invalid utm params which
+   * allows us to still submit the metrics.
+   *
+   * Coercion of data will be reported but NOT counted as a critical error.
+   */
+  public sanitizeUtmParams(data: CheckedUtmMetrics) {
+    const sanitize = (utmKey: keyof CheckedUtmMetrics) => {
+      const val = data[utmKey];
+      if (!this.isUtmValid(val)) {
+        // Report the error so we can see the issue, but do not consider it critical.
+        this.errorReporter.report(
+          new InvalidMetricError(
+            utmKey,
+            'invalid utm value',
+            false,
+            val,
+            INVALID_UTM
+          )
+        );
+
+        // Override original UTM param with `invalid` value. This will allow us
+        // to submit the metrics
+        data[utmKey] = INVALID_UTM;
+      }
+    };
+
+    sanitize('utm_campaign');
+    sanitize('utm_content');
+    sanitize('utm_medium');
+    sanitize('utm_term');
+    sanitize('utm_source');
+
+    return data;
+  }
+
+  /**
+   * Attempts to enforce a numeric range on a number
+   * @param {string} metricName - name of value being checked, if blank no error will be reported to sentry
+   * @param {any} val - value to clamp
+   * @param {number} min - minimum allowed value
+   * @param {number} max - maximum allowed value
+   * @param {any} def - default value in the event value is not a number
```

Comment:
Tiny nit, do you mind if this says `default value for if 'val' is not a number` or something slightly tweaked? I had to read this a couple times and read the return value to understand it, I think because of "in the event value" phrase.

Reply:
Thanks for pointing this out. That doc isn't very clear.